azdev/operations/performance.py

```diff
@@ -140,3 +144,103 @@ def display_table(data):
     for key, val in data.items():
         display('{:<20} {:>12.0f} {:>12.0f} {:>12.0f} {:>25}'.format(
             key, val['average'], val['threshold'], val['stdev'], str(val['values'])))
+
+
+# require azdev setup
+def benchmark(commands, runs=20):
+    if runs <= 0:
+        raise CLIError("Number of runs must be greater than 0.")
+
+    max_len_cmd = max(commands, key=len)
+
+    line_tmpl = "| {" + "cmd:" + "<" + str(len(max_len_cmd)) + "s} |"
+    line_tmpl = line_tmpl + " {min:10s} | {max:10s} | {avg:10s} | {mid:10s} | {std:10s} | {runs:10s} |"
+
+    line_head = line_tmpl.format(
+        cmd="Command",
+        min="Min",
+        max="Max",
+        avg="Mean",
+        mid="Median",
+        std="Std",
+        runs="Runs",
+    )
+
+    logger.warning(line_head)
+    logger.warning("-" * (85 + len(max_len_cmd)))
+
+    import multiprocessing
+
+    # Measure every wanted commands
+    for raw_command in commands:
+        logger.info("Measuring %s...", raw_command)
+
+        pool = multiprocessing.Pool(multiprocessing.cpu_count(), _benchmark_process_pool_init)
+        try:
+            time_series = pool.map_async(_benchmark_cmd_timer, [raw_command] * runs).get(1000)
+        except multiprocessing.TimeoutError:
+            pool.terminate()
+            break
+        else:
+            pool.close()
+        pool.join()
+
+        staticstic = _benchmark_cmd_staticstic(time_series)
+
+        line_body = line_tmpl.format(
+            cmd=raw_command,
+            min=str(staticstic["min"]),
+            max=str(staticstic["max"]),
+            avg=str(staticstic["avg"]),
+            mid=str(staticstic["media"]),
+            std=str(staticstic["std"]),
+            runs=str(runs),
+        )
+        logger.warning(line_body)
+
+    logger.warning("-" * (85 + len(max_len_cmd)))
+
+
+def _benchmark_process_pool_init():
+    import signal
+
+    def sigint_dummay_pass(signal_num, frame):  # pylint: disable=unused-argument
+        pass
+
+    signal.signal(signal.SIGINT, sigint_dummay_pass)
```

Comment:
Why do we need this function? Could you comment with some references?

Reply:
try/except like this because of a bug of Python multiprocessing.Pool https://bugs.python.org/issue8296

Doscissopm on StackOverflow: https://stackoverflow.com/questions/1408356/keyboard-interrupts-with-pythons-multiprocessing-pool/1408476

Otherwise, when user type ctrl+c to terminate the benchmark processing, it will left all the tedious exception stack on terminal which is un friendly.